### PR TITLE
Allow implementation and interface files to live in different dirs

### LINKED ISSUE #56: Virtual libs support
I was trying to set up a small example with dune [virtual libs](https://dune.readthedocs.io/en/stable/variants.html#virtual-library) but there seems to be an issue with bsb expectations. There is an interface file in the virtual lib folder and implementation in another (as expected), but build fails with the following, before even `dune.bsb` is generated:

```
$ esy x bsb -make-world
Error: Yesterday found in two directories: (js, shared_lib)
File names must be unique per project
```

Repro: https://github.com/jchavarri/hello-native-bucklescript/tree/074c06b162e2f0b9aa96f845a127a567c4c85985

```
    "@opam/bucklescript": "anmonteiro/bucklescript#1a00d85a81efb8f006d71a44222d1e3633d63d2e",
    "ocaml": "anmonteiro/ocaml#75f22c872451c66a4c4aadc43abee55697268a57"
```

### PR DESCRIPTION
**Pull request by anmonteiro**
### TL;DR

This change relaxes the restriction that implementation and interface files for the same module must live in the same directory.

### Motivation

- Dune's [virtual library](https://dune.readthedocs.io/en/stable/variants.html) support allows to compile parameterized libraries and defer the selection of a concrete implementation until linking.
  - It works by declaring a library with "virtual modules" (where only interface files are present), usually in a different directory than the corresponding implementation files
- This change allows projects that share code between native and JS targets can now rely on these interface files (in another directory), and satisfy those interfaces with their JS specific implementations, by relaxing the restriction that implementation and interface files for the same module must live in the same directory
- It makes code like [this example](https://github.com/jchavarri/hello-native-bucklescript/tree/virtual-lib) compile seamlessly

### Summary of the changes

- When processing "file groups", we try to merge groups for the same module rather than exiting with a hard failure
- We extend the `.bsbuild` database to account for modules where the implementation and interface files live in different directories
- We also allow having different extensions across implementations / interfaces (e.g. you can have Foo.re and Foo.mli)

fixes #56 

cc @jchavarri

### CODE REVIEW

**Comment by anmonteiro** [comment]
The Esy failures are related to the arm64 machine configuration.

This has been tested in a few projects and works nicely, so I'm gonna merge it to unblock other work.
